### PR TITLE
feat: use signed url instead of auth propagation

### PR DESCRIPTION
**Pull request by anda-ren**
## Description from #1422 

- use signed url instead of auth propagation in initContainers
- use signed url instead of auth propagation in client dataset
  - test for link raw type dataset http://sign-url-20221107.pre.intra.starwhale.ai/projects/3/evaluations/6/results
  - test for binary type dataset http://sign-url-20221107.pre.intra.starwhale.ai/projects/3/evaluations/7/results

## Modules
- [ ] UI
- [x] Controller
- [ ] Agent
- [x] Client
- [ ] Python-SDK
- [x] Others

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: client/starwhale/consts/__init__.py [hunk 1]
```diff
@@ -87,6 +87,7 @@ class SupportOS:
 class SWDSBackendType:
     S3 = "s3"
     LocalFS = "local_fs"
+    SignedUrl = "signed_url"
```
Comment (tianweidut) [inline client/starwhale/consts/__init__.py:90]: `signed_url` is part of `s3`?

File: client/starwhale/api/_impl/dataset/loader.py [hunk 2]
```diff
@@ -50,22 +49,28 @@ def _load_dataset_auth_env(self) -> None:
...
         self._stores[_k] = _store
         return _store
 
-    def _get_key_compose(self, row: TabularDatasetRow, store: ObjectStore) -> str:
+    def _get_key_compose(
+        self, row: TabularDatasetRow, store: ObjectStore
+    ) -> t.Tuple[str, int, int]:
         if row.object_store_type == ObjectStoreType.REMOTE:
-            data_uri = urlparse(row.data_uri).path
+            data_uri = row.data_uri
```
Comment (tianweidut) [inline client/starwhale/api/_impl/dataset/loader.py:73]:
```python
if row.object_store_type == ObjectStoreType.REMOTE:
            data_uri = row.data_uri
        else:
            data_uri = row.data_uri
```
-->
```python
data_uri = row.data_uri
if row.object_store_type != ObjectStoreType.REMOTE and store.key_prefix:
    data_uri = os.path.join(store.key_prefix, data_uri.lstrip("/"))
```
?

File: client/starwhale/api/_impl/dataset/loader.py [hunk 3]
```diff
@@ -50,22 +49,28 @@ def _load_dataset_auth_env(self) -> None:
             load_dotenv(auth_env_fpath)
 
     def _get_store(self, row: TabularDatasetRow) -> ObjectStore:
-        _k = f"{row.object_store_type.value}.{row.auth_name}"
+        _k = f"{self.dataset_uri}.{row.auth_name}"
```
Comment (tianweidut) [inline client/starwhale/api/_impl/dataset/loader.py:52]: different row can use different auth?

File: client/starwhale/core/dataset/store.py [hunk 4]
```diff
@@ -312,16 +322,18 @@ def from_data_link_uri(cls, data_uri: str, auth_name: str) -> ObjectStore:
...
-        if _type == InstanceType.STANDALONE:
-            backend = SWDSBackendType.LocalFS
-            bucket = str(DatasetStorage(dataset_uri).data_dir.absolute())
-        else:
-            backend = SWDSBackendType.S3
-            bucket = os.environ.get("SW_S3_BUCKET", _DEFAULT_S3_BUCKET)
-
-        return cls(backend=backend, bucket=bucket)
+    @classmethod
+    def from_remote_dataset(cls, dataset_uri: URI, auth_name: str) -> ObjectStore:
```
Comment (tianweidut) [inline client/starwhale/core/dataset/store.py:331]: what's different between [from_dataset_uri](https://github.com/star-whale/starwhale/pull/1443/files#diff-30bbfec45080e5cc83b6569a57f62aedfaffd66e819ce77222896ead95ce6c0dR322) and [from_remote_dataset](https://github.com/star-whale/starwhale/pull/1443/files#diff-30bbfec45080e5cc83b6569a57f62aedfaffd66e819ce77222896ead95ce6c0dR331)?

File: client/starwhale/core/dataset/store.py [hunk 5]
```diff
@@ -408,6 +415,30 @@ def _make_file(self, bucket: str, key_compose: str) -> FileLikeObj:
...
+        r = requests.get(url, headers=headers)
+        return io.BytesIO(r.content)
+
+    def sign_uri(self, uri: str, auth_name: str) -> t.Any:
+        r = self.do_http_request(
+            f"/project/{self.dataset_uri.project}/{self.dataset_uri.object.typ}/{self.dataset_uri.object.name}/version/{self.dataset_uri.object.version}/sign-link",
+            method=HTTPMethod.GET,
+            instance_uri=self.dataset_uri,
+            params={"uri": uri, "authName": auth_name, "expTimeMillis": 1000 * 60 * 30},
+        ).json()
```
Comment (tianweidut) [inline client/starwhale/core/dataset/store.py:438]: add [r.raise_for_status()](https://requests.readthedocs.io/en/latest/api/#requests.Response.raise_for_status) ?

File: client/starwhale/core/dataset/store.py [hunk 6]
```diff
@@ -408,6 +415,30 @@ def _make_file(self, bucket: str, key_compose: str) -> FileLikeObj:
...
+        _key, _start, _end = key_compose
+        url = self.sign_uri(_key, auth)
+        headers = {}
+        if _start and _end:
+            headers = {"Range": f"bytes={_start}-{_end}"}
+        r = requests.get(url, headers=headers)
+        return io.BytesIO(r.content)
+
+    def sign_uri(self, uri: str, auth_name: str) -> t.Any:
+        r = self.do_http_request(
```
Comment (tianweidut) [inline client/starwhale/core/dataset/store.py:433]: add timeout? requests default timeout is None, will blocking forever.

File: client/starwhale/core/dataset/store.py [hunk 7]
```diff
@@ -408,6 +415,30 @@ def _make_file(self, bucket: str, key_compose: str) -> FileLikeObj:
...
+    def _make_file(self, auth: str, key_compose: t.Tuple[str, int, int]) -> FileLikeObj:
+        _key, _start, _end = key_compose
+        url = self.sign_uri(_key, auth)
+        headers = {}
+        if _start and _end:
+            headers = {"Range": f"bytes={_start}-{_end}"}
+        r = requests.get(url, headers=headers)
+        return io.BytesIO(r.content)
+
+    def sign_uri(self, uri: str, auth_name: str) -> t.Any:
```
Comment (tianweidut) [inline client/starwhale/core/dataset/store.py:432]: add [http_retry](https://github.com/star-whale/starwhale/blob/main/client/starwhale/utils/retry.py#L33) decorator?

Comment (tianweidut) [inline client/starwhale/core/dataset/store.py:432]: return str ?

File: client/starwhale/core/dataset/store.py [hunk 8]
```diff
@@ -408,6 +415,30 @@ def _make_file(self, bucket: str, key_compose: str) -> FileLikeObj:
...
+        super().__init__(kind=SWDSBackendType.SignedUrl)
+        self.dataset_uri = dataset_uri
+
+    def _make_file(self, auth: str, key_compose: t.Tuple[str, int, int]) -> FileLikeObj:
+        _key, _start, _end = key_compose
+        url = self.sign_uri(_key, auth)
+        headers = {}
+        if _start and _end:
+            headers = {"Range": f"bytes={_start}-{_end}"}
+        r = requests.get(url, headers=headers)
```
Comment (tianweidut) [inline client/starwhale/core/dataset/store.py:429]: tune for file streaming read?

File: client/starwhale/core/dataset/store.py [hunk 9]
```diff
@@ -408,6 +415,30 @@ def _make_file(self, bucket: str, key_compose: str) -> FileLikeObj:
...
+class SignedUrlBackend(StorageBackend, CloudRequestMixed):
+    def __init__(self, dataset_uri: URI) -> None:
+        super().__init__(kind=SWDSBackendType.SignedUrl)
+        self.dataset_uri = dataset_uri
+
+    def _make_file(self, auth: str, key_compose: t.Tuple[str, int, int]) -> FileLikeObj:
+        _key, _start, _end = key_compose
+        url = self.sign_uri(_key, auth)
+        headers = {}
+        if _start and _end:
```
Comment (tianweidut) [inline client/starwhale/core/dataset/store.py:427]: if only set start or end, range does not work?

File: client/starwhale/core/dataset/store.py [hunk 10]
```diff
@@ -408,6 +415,30 @@ def _make_file(self, bucket: str, key_compose: str) -> FileLikeObj:
             return io.BytesIO(f.read(_end - _start + 1))
 
 
+class SignedUrlBackend(StorageBackend, CloudRequestMixed):
```
Comment (tianweidut) [inline client/starwhale/core/dataset/store.py:418]: add ut for `SignedUrlBackend` ?

File: client/starwhale/consts/__init__.py (hunk 1 above)
Comment (anda-ren) [reply client/starwhale/consts/__init__.py:92]: no. it's a new type

File: client/starwhale/core/dataset/store.py (hunk 10 above)
Comment (anda-ren) [reply client/starwhale/core/dataset/store.py:418]: done

File: client/starwhale/core/dataset/store.py (hunk 9 above)
Comment (anda-ren) [reply client/starwhale/core/dataset/store.py:427]: yes, both should be set

File: client/starwhale/core/dataset/store.py (hunk 8 above)
Comment (anda-ren) [reply client/starwhale/core/dataset/store.py:429]: let me do it in another pr

File: client/starwhale/core/dataset/store.py (hunk 7 above)
Comment (anda-ren) [reply client/starwhale/core/dataset/store.py]: done

Comment (anda-ren) [reply client/starwhale/core/dataset/store.py]: `self.do_http_request` has an [http_retry](https://github.com/star-whale/starwhale/blob/main/client/starwhale/utils/retry.py#L33) decorator

File: client/starwhale/core/dataset/store.py (hunk 6 above)
Comment (anda-ren) [reply client/starwhale/core/dataset/store.py:433]: `self.do_http_request` has default timeout scends

File: client/starwhale/core/dataset/store.py (hunk 5 above)
Comment (anda-ren) [reply client/starwhale/core/dataset/store.py:439]: done

File: client/starwhale/core/dataset/store.py (hunk 4 above)
Comment (anda-ren) [reply client/starwhale/core/dataset/store.py:331]: [from_remote_dataset](https://github.com/star-whale/starwhale/pull/1443/files#diff-30bbfec45080e5cc83b6569a57f62aedfaffd66e819ce77222896ead95ce6c0dR331) will build a signedUrlBackend, [from_dataset_uri](https://github.com/star-whale/starwhale/pull/1443/files#diff-30bbfec45080e5cc83b6569a57f62aedfaffd66e819ce77222896ead95ce6c0dR322) will build a boto3.s3 backend

File: client/starwhale/api/_impl/dataset/loader.py (hunk 3 above)
Comment (anda-ren) [reply client/starwhale/api/_impl/dataset/loader.py:57]: yes

File: client/starwhale/api/_impl/dataset/loader.py (hunk 2 above)
Comment (anda-ren) [reply client/starwhale/api/_impl/dataset/loader.py]: done

**Review by tianweidut — DISMISSED** [review]
LGTM

**Review by tianweidut — APPROVED** [review]

**Pull request merged**